File: algorithms/[211]添加与搜索单词 - 数据结构设计/solution.py
# topics = ["字典树", "深度优先搜索"]


from typing import Dict, Any


class WordDictionary:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = {}
        self.end = '#'

    def addWord(self, word: str) -> None:
        node = self.root
        for char in word:
            node = node.setdefault(char, {})
        node[self.end] = word

    def search(self, word: str) -> bool:
        def dfs(node: Dict[str, Any], s: str) -> bool:
            for idx, char in enumerate(s):
                if char == '.':
                    return any(dfs(node[key], s[idx + 1 :]) for key in node if key != self.end)
                if char not in node:
                    return False
                node = node[char]

            return self.end in node

        return dfs(self.root, word)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

File: algorithms/[211]添加与搜索单词 - 数据结构设计/solution.ts
// topics = ["设计", "字典树", "深度优先搜索"]

export class TrieNode {
  endOfWord: boolean;
  next: Map<string, TrieNode>;
  constructor() {
    this.endOfWord = false;
    this.next = new Map();
  }
}

class WordDictionary {
  root: TrieNode;
  constructor() {
    this.root = new TrieNode();
  }

  addWord(word: string): void {
    let node = this.root;

    for (const char of word) {
      if (!node.next.has(char)) {
        node.next.set(char, new TrieNode());
      }
      node = node.next.get(char)!;
    }

    node.endOfWord = true;
  }

  search(word: string): boolean {
    return this.find(word, this.root);
  }

  find(word: string, root: TrieNode): boolean {
    let node = root;

    for (let i = 0, len = word.length; i < len; i += 1) {
      const char = word[i];
      if (char == '.') {
        // 递归判断
        return [...node.next.values()].some((node) => this.find(word.slice(i + 1), node));
      }
      if (!node.next.has(char)) {
        return false;
      }
      node = node.next.get(char)!;
    }

    return node.endOfWord;
  }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */
